map-structure/vi-map/src/vi-map.cc

```diff
@@ -2971,34 +2970,44 @@ bool VIMap::mergeAllSubmapsFromMapWithoutResources(
               !base_last_vertex.getLandmarks().hasLandmark(submap_landmark_id));
 
           // If the same vertex in the submap owns a new landmark, we
-          // transfer it to the base map vertex.
-          vi_map::Landmark submap_landmark =
-              submap_first_vertex.getLandmarks().getLandmark(
-                  submap_landmark_id);
-
-          // Clear all observations and rebuild them later when adding the
-          // vertices.
-          base_last_vertex.getLandmarks().addLandmark(submap_landmark);
-          landmark_index.addLandmarkAndVertexReference(
-              submap_landmark_id, last_base_vertex_id);
+          // transfer it to the base map using the storing vertex unless
+          // the storing vertex is the first vertex of the submap,
+          // then we need to add it to the last vertex of the basemap.
           base_last_vertex.setObservedLandmarkId(
               frame_idx, observation_idx, submap_landmark_id);
-
-          // Check storing vertex of submap landmark after insertion into the
-          // base map.
-          CHECK(landmark_index.hasLandmark(submap_landmark_id));
-          const pose_graph::VertexId& storing_vertex_id_base_map =
-              landmark_index.getStoringVertexId(submap_landmark_id);
-          CHECK_EQ(storing_vertex_id_base_map, last_base_vertex_id);
-          CHECK(hasVertex(storing_vertex_id_base_map));
-          CHECK(getVertex(storing_vertex_id_base_map)
-                    .hasStoredLandmark(submap_landmark_id));
-          CHECK_GE(
-              base_last_vertex.getNumLandmarkObservations(submap_landmark_id),
-              1u);
-          CHECK(getLandmark(submap_landmark_id)
-                    .hasObservation(
-                        last_base_vertex_id, frame_idx, observation_idx));
+          const pose_graph::VertexId& new_landmark_vertex_id =
+              (storing_vertex_id_submap == first_submap_vertex_id)
+                  ? last_base_vertex_id
+                  : storing_vertex_id_submap;
+          CHECK(!landmark_index.hasLandmark(submap_landmark_id));
+          landmark_index.addLandmarkAndVertexReference(
+              submap_landmark_id, new_landmark_vertex_id);
+
+          // If the first vertex of the submap stores the landmark we also add
+          // it to the last vertex of the base map. Otherwise the landmark is
+          // stored in another vertex that will be added later.
+          if (storing_vertex_id_submap == first_submap_vertex_id) {
```

Comment:
we already know the outcome of this comparison from the assignment of `new_landmark_vertex_id`